chrome/content/zotero/contextPane.js

```diff
@@ -773,17 +760,140 @@ var ZoteroContextPane = new function () {
+		panelNotes.setAttribute('flex', '1');
+		panelNotes.setAttribute('orient', 'vertical');
+		var deck = document.createElement('deck');
+		deck.className = 'notes-deck';
+		deck.setAttribute('flex', '1');
+		panelNotes.append(deck);
+		var vbox2 = document.createElement('vbox');
+		var note = document.createElement('zoteronoteeditor');
+		note.setAttribute('flex', 1);
+		vbox2.append(note);
+		var vbox = document.createElement('vbox');
+		vbox.setAttribute('flex', '1');
+		vbox.setAttribute('class', 'zotero-box');
+		vbox.style.overflowY = 'auto';
+		panelNotes.append(vbox);
+		var hbox = document.createElement('hbox');
+		hbox.setAttribute('align', 'center');
+		var label = document.createElement('label');
+		var button = document.createElement('button');
+		button.hidden = readOnly;
+		button.setAttribute('label', Zotero.Intl.strings['zotero.item.add']);
+		button.addEventListener('click', () => {
+			deck.setAttribute('selectedIndex', 1);
+			var item = new Zotero.Item('note');
+			item.libraryID = parentItem.libraryID;
+			item.parentID = parentItem.id;
+			note.returnHandler = () => {
+				deck.setAttribute('selectedIndex', 0);
+				_updateAddToNote();
+			};
+			note.mode = editable ? 'edit' : 'view';
+			note.item = item;
+			note.focus();
+			_updateAddToNote();
+		});
+		hbox.append(label, button);
+		var grid = document.createElement('grid');
+		grid.setAttribute('flex', 1);
+		var columns = document.createElement('columns');
+		var column = document.createElement('column');
+		column.setAttribute('flex', 1);
+		columns.append(column);
+		var column = document.createElement('column');
+		columns.append(column);
+		grid.append(columns);
+		var rows = document.createElement('rows');
+		rows.setAttribute('flex', 1);
+		grid.append(rows);
+		vbox.append(hbox, grid);
+		deck.append(vbox, vbox2);
+		deck.setAttribute('selectedIndex', 0);
+		// Tags panel
+		var panelTags = document.createElement('tabpanel');
+		panelTags.setAttribute('orient', 'vertical');
+		panelTags.setAttribute('context', 'tags-context-menu');
+		panelTags.className = 'tags-pane';
+		panelTags.style.display = 'flex';
+		var div = document.createElementNS(HTML_NS, 'div');
+		div.className = 'tags-box-container';
+		div.style.display = 'flex';
+		div.style.flexGrow = '1';
+		panelTags.append(div);
+		var tagsBoxRef = React.createRef();
+		ReactDOM.render(
+			<TagsBoxContainer
+				key={'tagsBox-' + parentItem.id}
+				item={parentItem}
+				editable={!readOnly}
+				ref={tagsBoxRef}
+			/>,
+			div
+		);
+		// Related panel
+		var panelRelated = document.createElement('tabpanel');
+		var relatedBox = document.createElement('relatedbox');
+		relatedBox.setAttribute('flex', '1');
+		relatedBox.className = 'zotero-editpane-related';
+		panelRelated.addEventListener('click', (event) => {
+			if (event.originalTarget.closest('.zotero-clicky')) {
+				Zotero_Tabs.select('zotero-pane');
+			}
+		});
+		panelRelated.append(relatedBox);
```

Comment:
This is all recreating layout from elsewhere, right? This is going to get out of sync — if we're not going to create them together at this time (e.g., because the original layout is created in .xul files), we should add a bunch of "Keep in sync with…" comments.

Reply:
Yes, from itemPane.xul. Added a comment to keep things on both sides (itemPane.xul and contextPane.js) in sync.

The way how context pane works is a little bit hacky in many aspects. I expect that we extract item pane into a separate React container at some point and will be able to reuse it in context pane.